docs/proposals/20-ha-control-planes.md

```diff
@@ -0,0 +1,2614 @@
+
+#### Affinity and Anti-Affinity - `Preferred`
+
+Kubernetes provides `affinity/anti-affinity` rules on `nodes` and `pods` to influence the scheduling of pods of control plane components across failure domains. To influence the scheduling of pods across nodes/zones, `pod-affinity` and `pod-anti-affinity` rules can be used in combination.
+
+#### Topology Spread Constraints (a.k.a TSC)
+
+Topology spread constraints rely on node labels ([well-known-labels](https://kubernetes.io/docs/reference/labels-annotations-taints/)) to identify the topology domain(s) that each node is in. Please see [TSC](#references) documentation for further details. TSC provides finer control over scheduling of resources as compared to `affinity/anti-affinity` constraints.
+
+In order to spread replicas of a control plane component across several nodes/zones, the following TSC could be used:
+```yaml
+spec:
+  topologySpreadConstraints:
+  - maxSkew: 1
+    topologyKey: topology.kubernetes.io/zone
+    whenUnsatisfiable: DoNotSchedule
+    labelSelector:
+      matchLabels:
+        app: etcd-statefulset
+  - maxSkew: 1
+    topologyKey: kubernetes.io/hostname
+    whenUnsatisfiable: DoNotSchedule
+    labelSelector:
+      matchLabels:
+        app: <app-name>
+```
+> NOTE: One can combine `TopologySpreadConstraints` and `Affinity` rules.
+
+As per the documentation, the above two constrains will be evaluated in `conjunction`. In other words, these will be `and`ed together.
+The above constrains imply that, if there are more than 1 zone, then the first constraint will restrict 1 control plane component replica per zone (maxSkew=1) and the second constraint will be applied on nodes within a zone, further restricting one control plane component replica per node. This ideally covers HA scenarios for a `single-zone-multi-node` and `multi-zone-multi-node`.
+
+During testing of this feature, we found a few inconsistencies in the implementation and some quirks. See [appendix](#topology-spread-constraints-evaluation-and-findings) for findings.
+At the time of writing of this document, TSC is still not fully mature as is evident from the above findings. It is therefore the current recommendation to not use it for any scheduling decisions.
+
+## Gardener API Changes
+
+### Shoot Spec
+
+We propose to extend the `shoot` resource to have a dedicated field under `spec.kubernetes.controlplane` which allows the consumer to define `highAvailability` configuration.
+
+There are a few options that were considered:
+
+> NOTE: In the below mentioned HA configuration, the focus is on control plane HA. It is not clear what it would mean to define HA configuration for data plane components. However, for better clarity of scope and to provide extensibility in the future, HA configurations are housed under `controlPlane`.
+
+**Option #1:**
+```yaml
+kind: Shoot
+apiVersion: core.gardener.cloud/v1beta1
+spec:
+  kubernetes:
```

Comment:
Let's not move it below `kubernetes` but rather on top level (`.spec.controlPlane`, similar to `.spec.systemComponents`)